File: starterkit2/swift/TestControl.swift
class TestControl{

    // manager
    var m_curstate  : String? = nil
    var m_nextstate : String? = nil
    var m_noWait = false
    let MAX_CALL_STACK = 10
                                // 1   2   3   4   5   6   7   8   9   10
    var m_callstack : [String?] = [nil,nil,nil,nil,nil,nil,nil,nil,nil,nil]
    var m_callstack_level = 0

    func Update() {
        while true {
            m_noWait = false
            var bFirst = false
            if m_nextstate != nil {
                m_curstate = m_nextstate
                m_nextstate = nil
                bFirst = true
            }
            if  m_curstate != nil {
                switch m_curstate {
                    //[STATEGO OUTPUT START] indent(20) $/^S_/->#case$
                    //             psggConverterLib.dll converted from psgg-file:TestControl.psgg

                    case "S_END": S_END(bFirst)
                    case "S_GSB000": S_GSB000(bFirst)
                    case "S_HELLOWORLD": S_HELLOWORLD(bFirst)
                    case "S_HELLOWORLD2": S_HELLOWORLD2(bFirst)
                    case "S_HELLOWORLD3": S_HELLOWORLD3(bFirst)
                    case "S_HELLOWORLD4": S_HELLOWORLD4(bFirst)
                    case "S_HELLOWORLD5": S_HELLOWORLD5(bFirst)
                    case "S_LOP000": S_LOP000(bFirst)
                    case "S_LOP000_LoopCheckAndGosub____" : S_LOP000_LoopCheckAndGosub____(bFirst)
                    case "S_LOP000_LoopNext____" : S_LOP000_LoopNext____(bFirst)
                    case "S_NEWERA": S_NEWERA(bFirst)
                    case "S_RET000": S_RET000(bFirst)
                    case "S_RET001": S_RET001(bFirst)
                    case "S_SBS000": S_SBS000(bFirst)
                    case "S_SBS001": S_SBS001(bFirst)
                    case "S_START": S_START(bFirst)
                    case "S_SUBROUTINE": S_SUBROUTINE(bFirst)
                    case "S_SUBROUTINE1": S_SUBROUTINE1(bFirst)


                    //[STATEGO OUTPUT END]                    
                    default: print("Unexpected!")
                }
            }
            if !m_noWait { break }
        }
    }
    func Goto(_ st: String?) {
        m_nextstate = st;
    }
    func CheckState(_ st: String?) -> Bool {
        return m_curstate == st
    }
    func HasNextState() -> Bool {
        return m_nextstate != nil
    }
    func NoWait() {
        m_noWait = true
    }
    func GoSubState(_ sub: String?, next: String?) {
 	    if m_callstack_level >= MAX_CALL_STACK-1 {
		    print("CALL STACK OVER FLOW")
	    }
        else {
            m_callstack[m_callstack_level] = next;
            m_callstack_level += 1
            Goto(sub)      
        }
    }
    func ReturnState()
    {
	    var st: String? = nil;
        if m_callstack_level <= 0 {
		    print("CALL STACK UNDER FLOW")
	    }
        else {
            m_callstack_level -= 1
            st = m_callstack[m_callstack_level]
            Goto(st)
        }
    }

    // state
    //[STATEGO OUTPUT START] indent(4) $/./$
    //             psggConverterLib.dll converted from psgg-file:TestControl.psgg

    /*
        S_END
    */
    func S_END(_ bFirst : Bool ) {
         // end of state machine
    }
    /*
        S_GSB000
    */
    func S_GSB000(_ bFirst : Bool ) {
        GoSubState("S_SBS000", next: "S_LOP000")
        NoWait()
    }
    /*
        S_HELLOWORLD
    */
    func S_HELLOWORLD(_ bFirst : Bool ) {
        if bFirst {
            print("Hello World.")
        }
        if !HasNextState()
        {
            Goto("S_NEWERA");
        }
    }
    /*
        S_HELLOWORLD2
    */
    func S_HELLOWORLD2(_ bFirst : Bool ) {
        if bFirst {
            print("index = " + String(m_index) + " : result = First.")
        }
        if !HasNextState()
        {
            Goto("S_RET001");
        }
    }
    /*
        S_HELLOWORLD3
    */
    func S_HELLOWORLD3(_ bFirst : Bool ) {
        if bFirst {
            print("index = " + String(m_index) + " : result = Second.")
        }
        if !HasNextState()
        {
            Goto("S_RET001");
        }
    }
    /*
        S_HELLOWORLD4
    */
    func S_HELLOWORLD4(_ bFirst : Bool ) {
        if bFirst {
            print("index = " + String(m_index) + " : result = Third.")
        }
        if !HasNextState()
        {
            Goto("S_RET001");
        }
    }
    /*
        S_HELLOWORLD5
    */
    func S_HELLOWORLD5(_ bFirst : Bool ) {
        if bFirst {
            print("index = " + String(m_index) + " : result = " + String(m_val))
        }
        if !HasNextState()
        {
            Goto("S_RET001");
        }
    }
    /*
        S_LOP000
    */
    var m_index = 0
    func S_LOP000(_ bFirst : Bool )
    {
        m_index = 0
        Goto("S_LOP000_LoopCheckAndGosub____")
        NoWait()
    }
    func S_LOP000_LoopCheckAndGosub____(_ bFirst : Bool )
    {
        if m_index < 10 {
            GoSubState("S_SBS001", next: "S_LOP000_LoopNext____")
        }
        else {
            Goto("S_END")
        }
        NoWait()
    }
    func S_LOP000_LoopNext____(_ bFirst : Bool )
    {
        m_index += 1
        Goto("S_LOP000_LoopCheckAndGosub____")
        NoWait()
    }
    /*
        S_NEWERA
    */
    func S_NEWERA(_ bFirst : Bool ) {
        if bFirst {
            print("We are in the visual programming era.")
        }
        if !HasNextState()
        {
            Goto("S_GSB000");
        }
    }
    /*
        S_RET000
    */
    func S_RET000(_ bFirst : Bool ) {
        ReturnState()
        NoWait()
    }
    /*
        S_RET001
    */
    func S_RET001(_ bFirst : Bool ) {
        ReturnState()
        NoWait()
    }
    /*
        S_SBS000
    */
    func S_SBS000(_ bFirst : Bool ) {
        if !HasNextState()
        {
            Goto("S_SUBROUTINE");
        }
    }
    /*
        S_SBS001
    */
    func S_SBS001(_ bFirst : Bool ) {
        if !HasNextState()
        {
            Goto("S_SUBROUTINE1");
        }
    }
    /*
        S_START
    */
    func S_START(_ bFirst : Bool ) {
        Goto("S_HELLOWORLD")
        NoWait()
    }
    /*
        S_SUBROUTINE
    */
    func S_SUBROUTINE(_ bFirst : Bool ) {
        if bFirst {
            print("It is a subroutine.")
        }
        if !HasNextState()
        {
            Goto("S_RET000");
        }
    }
    /*
        S_SUBROUTINE1
    */
    var m_val = 0
    func S_SUBROUTINE1(_ bFirst : Bool ) {
        if bFirst {
            m_val = Int.random(in: 1..<8)
        }
        if  m_val == 1 { Goto( "S_HELLOWORLD2" ); }
        else if m_val == 2 { Goto( "S_HELLOWORLD3" ); }
        else if m_val == 3 { Goto( "S_HELLOWORLD4" ); }
        else { Goto( "S_HELLOWORLD5" ); }
    }


    //[STATEGO OUTPUT END]

    //Run
    func Run() -> Void {
        print("RUN!")
        Goto( "S_START" )
        while CheckState( "S_END" ) == false {
            Update()
        }
        print("END OF RUN")
    }
}
//let sm = TestControl()
//sm.Run()

File: starterkit2/ruby/TestControl.rb
# TestControl.rb
class TestControl

    MAX_CALL_STACK = 10

    def initialize
        @m_curfunc = nil
        @m_nextfunc = nil
        @m_bNoWait = false

                      # 1   2   3   4   5   6   7   8   9   10
        @m_callstack = [nil,nil,nil,nil,nil,nil,nil,nil,nil,nil]
        @m_callstack_level = 0
    end

    def update
        while true 
            bFirst = true
            if @m_nextfunc != nil then
                @m_curfunc = @m_nextfunc
                @m_nextfunc = nil
                bFirst = true
            end

            @m_bNoWait = false
            if @m_curfunc != nil then
                @m_curfunc.call(bFirst)            
            end

            if @m_bNoWait == false then
                break
            end
        end
    end

    def goto(st)
        @m_nextfunc = st
    end

    def checkState(st)
        return @m_curfunc == st
    end

    def noWait
        @m_bNoWait = true
    end

    def hasNextState
        return @m_nextfunc != nil
    end

    def goSubState(sub,nex)
        if @m_callstack_level >= (MAX_CALL_STACK-1) then
            puts('CALL STACK OVER FLOW')
        else
            @m_callstack[@m_callstack_level] = nex
            @m_callstack_level += 1
            goto( sub )
        end
    end

    def returnState
        st = nil
        if @m_callstack_level <= 0 then
            puts('CALL STACK UNDER FLOW')
        else
            @m_callstack_level -= 1
            st = @m_callstack[@m_callstack_level]
            goto( st )
        end
    end

    def run
        puts "[START OF RUNNING]"
        puts ""
        
        goto(self.method(:S_START))
        loopmax = 50000
        loopmax.times do |count|
            if count >= loopmax - 1 then
                puts "error : out of loop"
                break
            end
            update()
            if checkState(self.method(:S_END)) then
                puts ""
                puts "[END OF RUNNING]"
                break
            end
        end
    end

    # Embed code
    # [STATEGO OUTPUT START] indent(4) $/^E_/$
    #             psggConverterLib.dll converted from psgg-file:TestControl.psgg

    #
    #    E_0001
    #
    #
    TITLE="Ruby Sample using StateGo."


    # [STATEGO OUTPUT END]

    # State methods
    # [STATEGO OUTPUT START] indent(4) $/^S_/$
    #             psggConverterLib.dll converted from psgg-file:TestControl.psgg

    #
    #    S_1st
    #
    #
    def S_1st(bFirst)
        if bFirst then
            puts("1st")
        end
        if hasNextState() == false then
            goto(self.method(:S_RET001))
        end
    end
    #
    #    S_2nd
    #
    #
    def S_2nd(bFirst)
        if bFirst then
            puts( "2nd" )
        end
        if hasNextState() == false then
            goto(self.method(:S_RET001))
        end
    end
    #
    #    S_END
    #
    #
    def S_END(bFirst)
        # nothing
    end
    #
    #    S_GET_RANDOM
    #    乱数取得
    #
    def S_GET_RANDOM(bFirst)
        @x = rand(10)
        if @x == 0 then
            goto( self.method(:S_Zero) )
        elsif @x == 1 then
            goto( self.method(:S_1st) )
        elsif @x == 2 then
            goto( self.method(:S_2nd) )
        else
            goto( self.method(:S_Others) )
        end
    end
    #
    #    S_GSB000
    #
    #
    def S_GSB000(bFirst)
        goSubState(self.method(:S_SBS000), self.method(:S_PAS000))
        noWait()
    end
    #
    #    S_HELLOWORLD
    #
    #
    def S_HELLOWORLD(bFirst)
        if bFirst then
            puts("Hello World.")
        end
        if hasNextState() == false then
            goto(self.method(:S_HELLOWORLD1))
        end
    end
    #
    #    S_HELLOWORLD1
    #
    #
    def S_HELLOWORLD1(bFirst)
        if bFirst then
            puts(TITLE)
        end
        if hasNextState() == false then
            goto(self.method(:S_GSB000))
        end
    end
    #
    #    S_LOP000
    #
    #
    def S_LOP000(bFirst)
        @m_index = 0
        goto(self.method(:S_LOP000_cond____))
        noWait()
    end
    def S_LOP000_cond____(bFirst)
        if @m_index < 10 then
            goSubState(self.method(:S_SBS001),self.method(:S_LOP000_next____))
        else
            goto(self.method(:S_END))
        end
        noWait()
    end
    def S_LOP000_next____(bFirst)
        @m_index += 1
        goto(self.method(:S_LOP000_cond____));
        noWait()
    end
    #
    #    S_Others
    #
    #
    def S_Others(bFirst)
        if bFirst then
            puts(@x)
        end
        if hasNextState() == false then
            goto(self.method(:S_RET001))
        end
    end
    #
    #    S_PAS000
    #
    #
    def S_PAS000(bFirst)
        if hasNextState() == false then
            goto(self.method(:S_RAND_INIT))
        end
    end
    #
    #    S_RAND_INIT
    #    乱数初期化
    #
    def S_RAND_INIT(bFirst)
        if bFirst then
            srand(Time.now.tv_sec)
        end
        if hasNextState() == false then
            goto(self.method(:S_LOP000))
        end
    end
    #
    #    S_RET000
    #
    #
    def S_RET000(bFirst)
        returnState()
        noWait()
    end
    #
    #    S_RET001
    #
    #
    def S_RET001(bFirst)
        returnState()
        noWait()
    end
    #
    #    S_SBS000
    #
    #
    def S_SBS000(bFirst)
        goto(self.method(:S_SUBROUTINE))
        noWait()
    end
    #
    #    S_SBS001
    #
    #
    def S_SBS001(bFirst)
        goto(self.method(:S_START_LOOP))
        noWait()
    end
    #
    #    S_START
    #
    #
    def S_START(bFirst)
        goto(self.method(:S_HELLOWORLD))
        noWait()
    end
    #
    #    S_START_LOOP
    #
    #
    def S_START_LOOP(bFirst)
        if bFirst then
            print( "Loop :", @m_index, ", rand num =>" )
        end
        if hasNextState() == false then
            goto(self.method(:S_GET_RANDOM))
        end
    end
    #
    #    S_SUBROUTINE
    #
    #
    def S_SUBROUTINE(bFirst)
        if bFirst then
            puts('This is the subroutine.' )
        end
        if hasNextState() == false then
            goto(self.method(:S_RET000))
        end
    end
    #
    #    S_Zero
    #
    #
    def S_Zero(bFirst)
        if bFirst then
            puts("Zero")
        end
        if hasNextState() == false then
            goto(self.method(:S_RET001))
        end
    end


    # [STATEGO OUTPUT END]
end